Implement dynamic volume limits

gnufied opened the pull request:
Implement dynamic volume limits depending on node type.

xref https://github.com/kubernetes/community/pull/2051

```release-note
Add Alpha support for dynamic volume limits based on node type
```

gnufied committed:
Implement API changes needed for dynamic volume limits

define alpha feature and make api changes

gnufied committed:
Implement scheduler changes for volume limits

gnufied committed:
Implement volume plugin changes for volume limits

gnufied committed:
Implement kubelet side changes for writing volume limit to node

Add tests for checking node limits

gnufied commented:
/retest

jsafrane commented on pkg/volume/aws_ebs/aws_ebs.go:
```diff
@@ -93,6 +96,39 @@ func (plugin *awsElasticBlockStorePlugin) SupportsBulkVolumeVerification() bool
+
+	instanceType, err := instances.InstanceType(context.TODO(), plugin.host.GetNodeName())
+	if err != nil {
+		glog.V(3).Infof("Failed to get instance type from AWS cloud provider")
```
log the error

jsafrane commented on pkg/volume/plugins.go:
```diff
@@ -216,6 +216,20 @@ type ExpandableVolumePlugin interface {
+// volumes that can be attached to a node.
+type VolumePluginWithAttachLimits interface {
+	VolumePlugin
+	// Return number of volumes attachable on a node from the plugin
```
* Please add some description what is the return map - what's the key and what's value.
* I'd emphasize that it's called by kubelet and it returns limit for the node where this function is called.

jsafrane commented on pkg/volume/plugins.go:
```diff
@@ -216,6 +216,20 @@ type ExpandableVolumePlugin interface {
+	//    - storage-limits-aws-ebs
+	//    - storage-limits-csi-cinder
+	// The key should respect character limit of ResourceName type
+	VolumeLimitKey(spec *Spec) string
```
Add a note that this function is called by scheduler (or can be called by any Kubernetes component, not just kubelet).

gnufied commented on pkg/features/kube_features.go:
```diff
@@ -272,6 +272,13 @@ const (
+	//
+	// Add support for volume plugins to report node specific
+	// volume limits
+	DynamicVolumeThreshold utilfeature.Feature = "DynamicVolumeThreshold"
```
Limit is basically another word for threshold right? I just copied what was in specs. It should be pretty easy to rename the feature flag but I don't feel very strongly about this either ways.

jingxu97 replied:
Since we use "Limit" in other places, I will vote for "limit". Also agree on "AttachableVolume" part

derekwaynecarr commented:
i will help review kubelet changes.

/assign @derekwaynecarr

jingxu97 commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -415,6 +415,136 @@ func (c *MaxPDVolumeCountChecker) filterVolumes(volumes []v1.Volume, namespace s
 	return nil
 }
 
+func (c *MaxPDVolumeCountChecker) attachableLimitPredicate(
```
Add comment for this function?

gnufied replied:
fixed

jingxu97 commented:
cc @bsalamat Could you please help take a look at the scheduler change? Thanks!

gnufied commented:
For scheduler

/assign @bsalamat @aveshagarwal 

for api and other misc. changes

/assign @liggitt

gnufied commented:
/test pull-kubernetes-kubemark-e2e-gce-big
/test pull-kubernetes-e2e-kops-aws

liggitt commented on pkg/volume/plugins.go:
```diff
@@ -216,6 +216,32 @@ type ExpandableVolumePlugin interface {
+	//    - storage-limits-aws-ebs
+	//    - storage-limits-csi-cinder
+	// The key should respect character limit of ResourceName type
+	// This function may be called by kubelet or scheduler to identify node allocatable property
```
how does the scheduler call volume plugin methods?

gnufied replied:
Addressed below in response to a comment by @bsalamat  https://github.com/kubernetes/kubernetes/pull/64154#discussion_r190767908

gnufied replied:
fixed. removed volume plugin manager from scheduler. We are probably never going to support dynamic volume limits for flex. We will revisit the problem for CSI in 1.12.

bsalamat commented on pkg/scheduler/factory/factory.go:
```diff
@@ -186,6 +191,16 @@ func NewConfigFactory(
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.DynamicVolumeThreshold) {
+		volumePlugins := volumeplugins.ProbeVolumePlugins()
+		c.volumePluginMgr = &volume.VolumePluginMgr{}
```
Why do we need to add yet another volume plugin to the guts of the scheduler? I don't remember this was discussed in the design doc.

gnufied replied:
So this is not just another volume plugin. This brings in volume plugin manager, so as we can identify which volume type is being used in pod without hardcoding each volume plugin - https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/predicates/predicates.go#L493 . This also lets us identify which resource key will be present in node allocatable to check volume attach limit etc. 

I am going to update the design to reflect this. This was brought up during proposal discussions and I thought I addressed it...

gnufied replied:
I have explicitly called this out in design doc - https://github.com/gnufied/community/blob/055f8da12bbebca787445da4fcd7352996278712/contributors/design-proposals/storage/dynamic_volume_limit.md#changes-to-scheduler now

To some extent, I do not understand opposition to adding volume plugin manager in scheduler because alternative one way or another is to hardcode and switch..case based on volume type.

There are several ways we can workaround this, if it is a blocker:

1. We can hardcode for in-tree volume types in the predicate (Azure, AWS, GCE-PD and CSI). CSI will require special handling because `VolumeLimitKey` has to be generated by in-tree CSI plugin that exists in k8s.  There are two ways to fix CSI handling:
    - Rather than importing volume plugin manager in scheduler, we can just import CSI in-tree volume plugin. So we can hardcode type checks for aws, gce, azure and for CSI we can call `VolumeLimitKey` function defined inside CSI plugin to get the key. Pros of this approach is: we can reuse `VolumeLimitKey` logic between scheduler, kubelet and other components. Also we don't have to import volume plugin manager, because that imports all volume plugins.
    - Another option is - we can duplicate code inside `VolumeLimitKey` for CSI and copy it into scheduler. Problem with this approach is - overtime the `VolumeLimitKey` implemntation in volume plugin and scheduler may drift apart. :(

2. Another option @liggitt has proposed is https://github.com/kubernetes/community/pull/2051#discussion_r190778327 . What that does is - it adds a `ResourceRequirement` field to `CSIPersistentVolumeSource` with `{"storage-attach-request-aws-ebs": 1 } `.  There are couple of problems with this I think: 
    - Who decides key name inside `ResourceRequirement` map of `CSIPersistentVolumeSource`? For various reasons - it can't be same as CSI plugin name and plugin name is only string that uniquely identifies a CSI driver.  Also, CSI provisioner and plugin on node are two components, so we are still left with duplicating logic that will generate `VolumeLimitKey` in multiple places.
    - I can't help but think that only reason `CSIPersistentVolumeSource` carries `ResourceRequirement` is to carry key that will match resource name in node's allocatable for CSI.  So it is not a resource requirement of PV in typical sense.

gnufied replied:
@msau42 proposed another option that is, we hardcode volume types for AWS, GCE-PD, Azure and CSI. But for CSI we can write a util function that is used both in scheduler and CSI volume plugin to generate `VolumeLimitKey` . cc @liggitt do you think this works?

gnufied replied:
>The provisioner can be told what resource name to use in a storageclass parameter (as it does for fstype) or as a startup option (as it does with --volume-name-prefix) and set it on the PV when it constructs it here:

The resource name has to be available in volume plugin driver running on the node too(so as it can be registered in node's allocatable).  The problem is - a CSI plugin is CO agnostic component and it doesn't know anything about Kubernetes or resource name restriction that applies to names chosen by it.

> as those transition to CSI, I expected the CSI replacements for AWS/GCE/Azure to be able to continue indicating the same attach count resource name for the CSI PVs they provision. That implies the resource name is specified in the CSI PV, not derived from the CSI plugin or driver name

Yeah I definitely see value in this. However current in-tree to CSI migration plan is considering having CSI spec of a in-tree PV as a annotation in the PV itself. From scheduler perspective - this will allow us to consider a non-CSI volume as CSI and do the counting.

gnufied replied:
Remove volume plugin manager from scheduler

bsalamat commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -373,36 +388,21 @@ func (c *MaxPDVolumeCountChecker) filterVolumes(volumes []v1.Volume, namespace s
 				return fmt.Errorf("PersistentVolumeClaim had no name")
 			}
 
-			// Until we know real ID of the volume use namespace/pvcName as substitute
```
Is the removal of these parts related to the enforcement of volume limits? If not, please don't make the changes in this PR.

gnufied replied:
removal of these bits from existing predicates was because this code path no longer applies. Users can not delete in-use PVCs and users can not delete PV bound to PVCs. I agree that, perhaps this can go in separate PR but removing them here helps me in using same unit tests for both old and new predicates. 

I can move this to separate PR (or do nothing) - in which case I will have to carefully duplicate unit tests, because I do not want old behavior (of safeguarding against deleted pvcs/pv) in new predicate.

gnufied replied:
reverted these bits. We will revisit this later.

gnufied commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -346,6 +350,17 @@ func NewMaxPDVolumeCountPredicate(filterName string, pvInfo PersistentVolumeInfo
 }
 
+// NewMaxAttachableVolumeLimitPredicate returns a predicate that can count maximum number of attachable volumes
+func NewMaxAttachableVolumeLimitPredicate(
```
This feature being Alpha we can't perhaps remove existing predicates, but we can document how user can use just this predicate via scheduler policy file that lists this predicate but drops others. Alternately when alpha feature gate is enabled and volume limits are available from node objects, we can make other predicates NOOP.

aveshagarwal replied:
I think when this feature is enabled you could do something like this: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithmprovider/defaults/defaults.go#L177-#L180 to disable those old predicates.

gnufied replied:
I think we may not want to remove the old predicates if feature gate is enabled because it is possible that schduler is running on 1.11 and with feature gate but nodes do not have feature gate or is running with older version.

I will make other predicates instead NOOP if feature gate is enabled and node's allocatable property contains volume limits.

aveshagarwal replied:
Also wondering, why not just use the existing predicates instead of creating a new one? My understanding after discussion with you was that: 1) volumelimits are exposed via kubelet in allocatable that would be used by the scheduler . 2) so for these existing predicate instead of using current hardcoded limits, the dynamically limits set in step 1 would be used.

gnufied replied:
There are three existing predicates `MaxEBSVolumeCountPred` , `MaxGCEPDVolumeCountPred` and `MaxAzureDiskVolumeCountPred` which work only for those 3 volume types (EBS, GCE-PD and Azure).

This PR presents a design which will make VolumeAttachableLimit predicate work for ANY volume type, as long as they implement necessary interface.  

If we updated existing predicate, it will still only only to those 3 volume types and won't work for new volume types like CSI or anything else.

liggitt replied:
> My understanding after discussion with you was that: 1) volumelimits are exposed via kubelet in allocatable that would be used by the scheduler . 2) so for these existing predicate instead of using current hardcoded limits, the dynamically limits set in step 1 would be used.

if the scheduler was making use of these resources in the same way as existing resources that would work. the reason that doesn't work here is that nothing in the API indicates that resource is consumed... the scheduler has to infer it from the volume types

aveshagarwal replied:
To keep the integration between volumeManager and scheduler low, would this flow work?: 1) have an admission plugin  that adds (integer) resources request to an input pod based on volume types in the pod spec by calling volumeManager 2) scheduler just works based on those added resources requests and dynamically added limits via kubelet. Overall It would also help with avoiding any overhead (if any) related to calling volumeManager.  Just throwing it as an idea.

gnufied replied:
Yeah that was discussed and we chose not to implement that because:

a. We will have to make an API change to pod spec to store such a counted resource somewhere. It was discussed that, if we made such a API change - it will be probably only used for counting volumes and hence it is bit premature to add such a field to pod spec. If there were other resources that needed to be counted at pod level, we can consider the API change.  This proposal however does not excludes taking this route if we decide in future. in Beta for example - we can certainly do this, if that is path we want to take.
b. There was already a precedence for using resources that are available in allocatable but not explicitly requestable by pods (max pod count). sorry @liggitt I forgot about `pod` field in allocatable, when we discussed this yesterday. :( 

Also - I think we can remove integration between volume plugin manager and scheduler, I responded this https://github.com/kubernetes/kubernetes/pull/64154#discussion_r190955132 . I will update the proposal.

liggitt replied:
> We will have to make an API change to pod spec to store such a counted resource somewhere. It was discussed that, if we made such a API change - it will be probably only used for counting volumes and hence it is bit premature to add such a field to pod spec.

pinned resource sets for the in-tree volume use seems ok. I'm much more interested in the representation of resource consumption (including attach count consumption) for CSI volumes, which are only in PV objects, not in the pod spec.

gnufied replied:
Fixed the original comment. This PR no longer introduces a new predicate. Because this PR only handles GCE, AWS and Azure and there are already predictes for that, this PR merely changes source of max Volume count.

aveshagarwal commented on pkg/kubelet/kubelet_test.go:
```diff
@@ -133,13 +136,30 @@ func newTestKubelet(t *testing.T, controllerAttachDetachEnabled bool) *TestKubel
+	return newTestKubeletWithImageList(t, imageList, controllerAttachDetachEnabled, true /*initFakeVolumePlugin*/)
+}
+
+func newTestKubeletWithouFakeVolumePlugin(t *testing.T, controllerAttachDetachEnabled bool) *TestKubelet {
```
nit: seems like you missed `t` in `Withou`.

aveshagarwal commented on pkg/kubelet/kubelet_node_status_test.go:
```diff
@@ -1580,3 +1580,64 @@ func TestValidateNodeIPParam(t *testing.T) {
+		kubelet.cloudproviderRequestSync = make(chan int)
+		kubelet.cloudproviderRequestTimeout = 10 * time.Second
+		kubelet.setVolumeLimits(node)
+		volumeLimits := node.Status.Allocatable
```
test capacity too.

gnufied replied:
fixed.

aveshagarwal commented on pkg/scheduler/algorithm/predicates/max_attachable_predicate_test.go:
```diff
@@ -0,0 +1,866 @@
+/*
```
i think the name of this file as max_attachable_volume_predicate_test.go would be more helpful.

gnufied replied:
fixed

aveshagarwal commented on pkg/scheduler/algorithm/predicates/max_attachable_predicate_test.go:
```diff
@@ -0,0 +1,866 @@
+				},
+			},
+		},
+	}
```
could you use a function to create all various pods by passing required arguments?

gnufied replied:
I just extracted what was already there in separate file, so as all volume count predicate related tests can be in one file.

aveshagarwal commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -289,10 +292,11 @@ func NoDiskConflict(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *sch
+	maxVolumes      int
+	pvInfo          PersistentVolumeInfo
+	pvcInfo         PersistentVolumeClaimInfo
+	volumePluginMgr *volume.VolumePluginMgr
```
May be I am missing something but when at the time of node registration, these volume limits are already published in capacity and allocatable, why do you need to expose this again in scheduler?

gnufied replied:
Currently the scheduler uses `volumePluginMgr` interface to identify which type of volume plugin the pod is using and what would be `VolumeLimitKey` for the plugin, so as scheduler can match that value with key/value present in node's Allocatable.

msau42 replied:
We don't need volumePluginMgr anymore right?

gnufied commented:
@liggitt @saad-ali @bsalamat @aveshagarwal addressed most review comments in code. Dropped volume plugin manager. This PR also no longer introduces a new predicate, because this PR in scheduler merely changes source of "maxVolumeLimit". Earlier it was hardcoded in scheduler and now if feature gate is enabled, will be picked from node's allocatable.

I will update design doc too, soonish.

gnufied commented:
/retest

gnufied commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

jsafrane commented:
volume plugins lgtm
(and the rest does not look bad either)

gnufied commented:
/kind/feature
/priority/critical-urgent

jsafrane commented:
/kind feature
/priority critical-urgent

msau42 commented on pkg/features/kube_features.go:
```diff
@@ -272,6 +272,13 @@ const (
 	VolumeSubpath utilfeature.Feature = "VolumeSubpath"
 
+	// owner: @gnufied
+	// alpha
```
Add v1.11

gnufied replied:
fixed

msau42 commented on pkg/scheduler/factory/factory.go:
```diff
@@ -634,6 +634,9 @@ func (c *configFactory) updatePodInCache(oldObj, newObj interface{}) {
 		return
 	}
 
+	// NOTE: Because the scheduler uses snapshots of schedulerCache and the live
```
Are these your changes?

gnufied replied:
fixed

gnufied commented:
/retest

msau42 commented:
lgtm

saad-ali commented:
/approve

liggitt commented on pkg/apis/core/types.go:
```diff
@@ -3590,6 +3590,8 @@ const (
 	// Name prefix for huge page resources (alpha).
 	ResourceHugePagesPrefix = "hugepages-"
+	// Name prefix for storage resource limits
+	ResourceStorageLimitPrefix = "storage-attach-limits-"
```
I apologize if this is bikeshedding on the name, but this still reads oddly compared to other resources.

thinking of a node declaring it has a plural count of these resources, and storageclass/pv objects eventually declaring they consume one or more of these resources, this is how this would read:

* node: "capacity: {storage-attach-limits-aws-ebs: 39}"
* pv: "resources: {storage-attach-limits-aws-ebs: 1}"

putting "limits" in the name of the resource is odd... limits are inherent in it being a counted resource. would "attachable-volumes-..." be a better prefix?

* node: "capacity: {attachable-volumes-aws-ebs: 39}"
* pv: "resources: {attachable-volumes-aws-ebs: 1}"

@saad-ali @thockin, thoughts?

gnufied replied:
yeah. naming is hard. I can change it to `attachable-volumes-`  prefix

gnufied replied:
fixed.

gnufied commented:
@liggitt addressed your naming concern. PTAL

liggitt commented:
API resource prefix and negative validation test LGTM

/approve

msau42 commented:
/lgtm

gnufied commented:
/test pull-kubernetes-e2e-gce

bsalamat commented on pkg/apis/core/v1/helper/helpers.go:
```diff
@@ -92,10 +92,14 @@ func IsOvercommitAllowed(name v1.ResourceName) bool {
 		!IsHugePageResourceName(name)
 }
 
+func IsStorageLimitResourceName(name v1.ResourceName) bool {
```
I would rename this to `IsVolumeLimitResourceName` to be consistent with the rest of the code.

bsalamat reviewed (commented):
Scheduler changes look good to me. Just a minor comment.

msau42 commented:
/lgtm

bsalamat reviewed (commented):
/lgtm

for scheduler changes

andyzhangx commented on pkg/volume/azure_dd/azure_dd.go:
```diff
@@ -106,6 +111,22 @@ func (plugin *azureDataDiskPlugin) SupportsBulkVolumeVerification() bool {
+		return nil, fmt.Errorf("Expected Azure cloudprovider, got %s", cloud.ProviderName())
+	}
+
+	volumeLimits := map[string]int64{
```
Question: @gnufied it looks like I need to write code here to query volume limits according to azure node type , and assign new value to `volumeLimits ` in `azure_dd.go`, right?

gnufied replied:
yes that is correct. I did not know how we can get those limits for Azure Disk and hence I went with existing defaults.

yujuhong commented:
This broke the node e2e alpha suite: https://k8s-testgrid.appspot.com/sig-node-kubelet#kubelet-alpha

gnufied commented:
@yuanying looking. ty
